src/workerd/api/sql-test.js

```diff
@@ -322,6 +323,119 @@ async function test(sql) {
   assert.equal(sql.voluntarySizeLimit, 1073741823 * 4096);
   sql.voluntarySizeLimit = 65536;
   assert.equal(sql.voluntarySizeLimit, 65536);
+
+  storage.put("txnTest", 0);
+
+  // Try a transaction while no implicit transaction is open.
+  await scheduler.wait(1);  // finish implicit txn
+  let txnResult = await storage.transaction(async () => {
+    storage.put("txnTest", 1);
+    assert.equal(await storage.get("txnTest"), 1);
+    return "foo";
+  });
+  assert.equal(await storage.get("txnTest"), 1);
+  assert.equal(txnResult, "foo");
```

Comment:
What is the expected behavior if I am waiting on multiple transactions concurrently? A test should be added for that case:

```js
const txn1 = storage.transaction(async () => { /*... */ });
const txn2 = storage.transaction(async () => { /*... */})
await Promise.all([txn1, txn2]);  // or Promise.allSettled
```

Reply:
This is extensively covered in our internal test suite. Input gates prevent the transactions from overlapping -- txn1 will run to completion first, then txn2 will start.

Reply:
Note that this PR doesn't change the existing logic that accomplishes this.